enhancements/machine-config/machine-config-node.md

```diff
@@ -0,0 +1,340 @@
+```
+
+Where each name represents a node in a given machineconfigpool. The statuses reported are created explicitly from MCO node annotations and MCO actions, no other operator actions are taken into account here. This allows us to get quite specific in what is occuring on the nodes.
+
+```console
+Name:         ip-10-0-12-194.ec2.internal
+Namespace:    
+Labels:       <none>
+Annotations:  <none>
+API Version:  machineconfiguration.openshift.io/v1alpha1
+Kind:         MachineConfigNode
+Metadata:
+  Creation Timestamp:  2023-10-17T13:08:58Z
+  Generation:          1
+  Resource Version:    49443
+  UID:                 4bd758ab-2187-413c-ac42-882e61761b1d
+Spec:
+  Node Ref:
+    API Version:  
+    Kind:         Node
+    Name:         ip-10-0-12-194.ec2.internal
+  Pool:
+    API Version:  machineconfiguration.openshift.io/v1
+    Kind:         MachineConfigPool
+    Name:         master
+Status:
+  Conditions:
+    Last Transition Time:  2023-10-17T13:09:02Z
+    Message:               Node has completed update to config rendered-master-cf99e619747ab19165f11e3546c71f1e
+    Reason:                NodeUpgradeComplete
+    Status:                True
+    Type:                  Updated
+    Last Transition Time:  2023-10-17T13:09:02Z
+    Message:               This node has not yet entered the UpdatePreparing phase
+    Reason:                NotYetOccured
+    Status:                False
+  Config Version:
+    Current:            rendered-worker-823ff8dc2b33bf444709ed7cd2b9855b
+    Desired:            rendered-worker-823ff8dc2b33bf444709ed7cd2b9855b
+  Health:               Healthy
+  Most Recent Error:    
+  Observed Generation:  3
+```
+
+The above struct gives us some helpful information about a node as it pertains to the MCO. First, all upgrade related events that have occured on the node from the most recent uprade process (no matter how small) appear in the `conditions`. Other than that, you can see the current and desired machineconfig indicating whether or not the node should be updating as well as whether or not the currently tracked update process held in `conditions` is updating to the expected MachineConfig. The Node refernece and ObservedGeneration exist to let the user know some extra information on the object and how many times we have gone throgh some upgrade related changes.
+
+The states to be reported by this MachineConfigNode will roughly fall into the following:
+
+#### Preparing phase
+ - stopping config drift monitor
+ - reconciling configs
+#### Actionable phase
+- Draining
+- Updating on disk state
+- Updating OS
+#### Post update action
+- Rebooting
+- Closing daemon
+- Node Reboot **OR** Reloading crio
+
+This loops until we are inDesiredConfig
+
+More Paths that will update the node state:
+
+#### Completing Update -- the daemon only gets into this phase (outside of hypershift) when we are inDesiredConfig,it has to do with the drainer annotation
+- Uncordoning
```

Comment:
Unless this fails, this category of operations is quite fast. Could this be folded into `updated` instead?

Similarly with cordoningnode above. Drains are the heavy hitter, so maybe we can fold those conditions together?

Not opposed to what you have right now, just a thought

Reply:
One issue with trying to represent a real-world state machine with conditions is that there is tension between talking about what has been accomplished (e.g. "draining is done!") and what is being worked on now (e.g. "we are in the process of draining, and here's how it's going...").  For cordoning, it feels like there is one state angle (is the node currently cordoned?) and a few transitional aspects:

* We expect it to be cordoned in the future, but are waiting on something (e.g. the machine-config controller to bump our desired MachineConfig to ask us to update with an update that requires a drain).
* We are ready to cordon, but the cordoning hasn't been completed yet.  Might be so brief that it's not worth talking about.
* We are cordoned and expect to stay that way while other work progresses (e.g. draining, rebooting, etc.).
* We are ready to uncordon, but the uncordoning hasn't been completed yet.  Again, this is a brief thing, and it's hard to imaging it breaking except via paths like "we can't reach the Kube API server" that would also block the controller's ability to talk about the issue in a Kube API resource like the MachineConfigNode.
* We are uncordoned, and not aware of any pending situations that would cause us to re-cordon.

And I guess maybe also:

* The Node is uncordoned but feel like we should be cordoned at this stage, and is maybe somebody fighting with us over that Node knob?  Not sure what's going on.
* The Node is cordoned, and while we don't have a reason to need that ourselves, other actors are welcome to cordon the node for their own reasons.

Having conditions per-transition ends up being very chatty, especially if you decide [you need to have opinions on all possible conditions all the time][1].  Having one condition with `type` `Cordoned` with the usual `True`, `False`, and `Unknown` `status` values and a bunch of `reason`s for the various contexts is more compact.  But for `-o wide` rendering, are the status and reason sufficient context?

For example, pivoting to draining, the actual state is "are their drainable pods on this node?".  Would it be sufficiently expressive to have a `DrainablePods` type that flowed through:

| `status` | `reason` |
|-|-|
| `Unknown` | `DrainNotRequested` |
| `False` | `Draining` |
| `False` | `DrainingForALongTimePleaseHelp` :wink:|
| `True` | `Drained` (with an "I think?  Haven't actually checked in the past 10m" caveat?  Or just assuming that the pre-uncordon next-steps complete fast enough tht it's not worth caveating?) |

and eventually transitioned back to `Unknown`/`DrainNotRequested`?

[1]: https://github.com/openshift/enhancements/pull/1490/#discussion_r1372386774

Reply:
From a update point of view, understanding why `uncordoning` or `cordoning` getting stuck is going to help us uncovering some of the issues where updates take more time than expected.

Reply:
> From an update point of view, understanding why uncordoning or cordoning getting stuck is going to help us uncovering some of the issues where updates take more time than expected.

But you could get that from a `Cordoned` condition right, so long as the message contained any errors that were coming in to play?

I like what @wking is saying here, perhaps we can revisit the conditions that are being made in this API and take a look at what really needs to be said. Cordoning is a single operation that should, in theory, always work, so I'm leaning towards we should represent whether the node is cordoned, vs whether we are attempting to cordon it.

Reply:
@wking are you suggesting that we think of these actions as their results rather than the processes themselves? and for certain ones like drained and cordoned have the t/f/unknown map to specific reasons for what is happening when doing `oc describe mcn/<name>` ?

Reply:
I think that is the general use case.

To give an example for the updated case
```
type: Updated
status: False
reason: MachineConfigPoolVersionMismatch
message: MachineConfigPool desired version does not match Node desired version
```
and then
```
type: Updated
status: True
reason: MachineConfigPoolVersionMatch
message: Node has reached desired version from MachineConfigPool
```

Or for something like the files applied
```
type: AppliedOSAndFiles
status: False
reason: UpgradeIdle
message: No upgrade currently in progress
```
You could also have that state represented as
```
type: AppliedOSAndFiles
status: True
Reason: UpgradeApplied
message: OS and Files are applied to current desired version
```

Which begs the question, for some of these states, we need to define which would be the least surprising idle state?

If I'm an end user and I print this out, and see, Updated: True, then I'm probably happy and don't need to look further, but, if I see Updated: True, would I also expect to see things like AppliedOSAndFiles to be true? That way when the upgrade starts, and Updated flicks to false, AppliedOSAndFiles would also flick to false, rather than having already been idle at false?

I think I lean to the latter of those two examples personally.

In general when having discussion like this, it's good to find existing examples and see what others have done in this space, what would be a similar component 🤔